src/template.js

```diff
@@ -120,114 +109,199 @@ export class BaseTemplate {
+    return templateElement;
   }
 
-  const type = assert(element.getAttribute('type'),
-      'Type must be specified: %s', element);
+  /**
+   * Returns the promise that will eventually yield the template implementation
+   * for the specified template element.
+   * @param {!Element} element
+   * @return {!Promise<!BaseTemplate>}
+   * @private
+   */
+  getImplementation_(element) {
+    const impl = element[PROP_];
+    if (impl) {
+      return Promise.resolve(impl);
+    }
+
+    const type = assert(element.getAttribute('type'),
+        'Type must be specified: %s', element);
+
+    let promise = element[PROP_PROMISE_];
+    if (promise) {
+      return promise;
+    }
 
-  let promise = element[PROP_PROMISE_];
-  if (promise) {
+    promise = this.waitForTemplateClass_(element, type).then(templateClass => {
+      const impl = element[PROP_] = new templateClass(element);
+      delete element[PROP_PROMISE_];
+      return impl;
+    });
+    element[PROP_PROMISE_] = promise;
     return promise;
   }
 
-  promise = waitForTemplateClass(element, type).then(templateClass => {
-    const impl = element[PROP_] = new templateClass(element);
-    delete element[PROP_PROMISE_];
-    return impl;
-  });
-  element[PROP_PROMISE_] = promise;
-  return promise;
-}
-
+  /**
+   * Returns the promise that will eventually yield the template class. This will
+   * wait until the actual template script has been downloaded and parsed.
+   * @param {!Element} element
+   * @param {string} type
+   * @return {!Promise<!TemplateClass>}
+   * @private
+   */
+  waitForTemplateClass_(element, type) {
+    if (this.templateClassMap_[type]) {
+      return this.templateClassMap_[type];
+    }
 
-/**
- * Returns the promise that will eventually yield the template class. This will
- * wait until the actual template script has been downloaded and parsed.
- * @param {!Element} element
- * @param {string} type
- * @return {!Promise<!TemplateClass>}
- */
-function waitForTemplateClass(element, type) {
-  if (templateClassMap[type]) {
-    return templateClassMap[type];
+    this.checkTemplateDeclared_(element, type);
+    let aResolve;
+    const promise = new Promise((resolve, reject) => {
+      aResolve = resolve;
+    });
+    this.templateClassMap_[type] = promise;
+    this.templateClassResolvers_[type] = aResolve;
+    return promise;
   }
 
-  checkTemplateDeclared(element, type);
-  let aResolve;
-  const promise = new Promise((resolve, reject) => {
-    aResolve = resolve;
-  });
-  templateClassMap[type] = promise;
-  templateClassResolvers[type] = aResolve;
-  return promise;
-}
 
+  /**
+   * Checks that the template type has actually been declared by a
+   * `<script custom-template=$type>` tag in the head.
+   * @param {!Element} element
+   * @param {string} type
+   * @private
+   */
+  checkTemplateDeclared_(element, type) {
+    const scriptTags = this.win_.document.querySelectorAll(
```

Comment:
Ok. Will do. It's helpful, but not super helpful because this path gets visited only before the template gets to call registerTemplate. So it's a relatively short time when this can happen.